ktor-client/ktor-client-features/ktor-client-json/ktor-client-json-jvm/src/io/ktor/client/features/json/serializer/SerialHelperJvm.kt

```diff
@@ -0,0 +1,45 @@
+package io.ktor.client.features.json.serializer
+
+import io.ktor.client.call.TypeInfo
+import kotlinx.serialization.*
+import kotlinx.serialization.internal.*
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+import java.lang.reflect.WildcardType
+import kotlin.reflect.KClass
+
+actual fun serializerForPlatform(type: TypeInfo): KSerializer<*> = serializerByTypeToken(type.reifiedType)
+
+@Suppress("UNCHECKED_CAST")
+fun serializerByTypeToken(type: Type): KSerializer<Any> = when (type) {
+    is Class<*> -> if (!type.isArray) {
+        serializerByClass(type.kotlin)
+    } else {
+        val eType: Class<*> = type.componentType
+        val s = kotlinx.serialization.serializerByTypeToken(eType)
+        ReferenceArraySerializer<Any, Any>(eType.kotlin as KClass<Any>, s) as KSerializer<Any>
+    }
+    is ParameterizedType -> {
+        val rootClass = (type.rawType as Class<*>)
+        val args = type.actualTypeArguments.map { argument ->
+            when (argument) {
+                is WildcardType -> argument.upperBounds.first() as Class<*>
+                else -> argument
+            }
+        }
+        when {
+            List::class.java.isAssignableFrom(rootClass) -> ArrayListSerializer(serializerByTypeToken(args[0])) as KSerializer<Any>
```

Comment:
Could we introduce the extension point here to add user mappers?

Reply:
as a user i would register custom serializers on the SerialContext when setting
```
serializer = kotlinxSerializer() { 
    registerSerializer(...)
}
```
or by passing it a existing `JSON`
mappers as they were used were quite limited since they only mapped based on class and generics were erased
is there any usecase where mappers would be prefered ?

Reply:
we could make a extension point to allow users to map custom serializers though
backed by a `HasMap<Class, (List<Type>) -> KSerializer<Any>>` maybe ?

Reply:
Something that looks like:
`registerSerializer(Type) { /* Serializer */ } `

And we could store serializers in `List` instead of the hashmap to register multiple serializers and find first `not failing`.

Reply:
Sure. Let's wait for the PR and just upgrade serialization version.

Reply:
with `not failing`do you meant to try all serializers for a class and use the result of the first one that does not throw ?